solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseCloudSolrClient.java

```diff
@@ -580,23 +644,71 @@ public void registerDocCollectionWatcher(String collection, DocCollectionWatcher
           throw getRouteException(SolrException.ErrorCode.SERVER_ERROR, exceptions, routes);
         }
       }
-    } else {
-      for (Map.Entry<String, ? extends LBSolrClient.Req> entry : routes.entrySet()) {
-        String url = entry.getKey();
-        LBSolrClient.Req lbRequest = entry.getValue();
+      return null;
+    });
+    updateFuture.exceptionally((error) -> {
+      if (updateFuture.isCancelled()) {
+        for (CompletableFuture<NamedList<Object>> cf : futuresArray) {
+          cf.cancel(true);
+        }
+      }
+      return null;
+    });
+    return updateFuture;
+  }
+
+  private CompletableFuture<Void> doUpdatesWithoutExecutor(final Map<String, ? extends LBSolrClient.Req> routes,
```

Comment:
Since the asyc & sync code paths are rather different inside here, I think you should structure this method as a big if-else.  The only part in common is like 2 lines.